core/src/main/java/jenkins/slaves/RemotingVersionInfo.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2018, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.slaves;
+
+import hudson.util.VersionNumber;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+// TODO: Make the API public (JENKINS-48766)
+/**
+ * Provides information about Remoting versions used withing the core.
+ * @author Oleg Nenashev
+ * @since TODO
+ */
+@Restricted(NoExternalUse.class)
+public class RemotingVersionInfo {
+
+    private static final Logger LOGGER = Logger.getLogger(RemotingVersionInfo.class.getName());
+    private static final String RESOURCE_NAME="remoting-info.properties";
+
+    @CheckForNull
+    private static VersionNumber EMBEDDED_VERSION;
+
+    @CheckForNull
+    private static VersionNumber MINIMAL_SUPPORTED_VERSION;
+
+    private RemotingVersionInfo() {}
+
+    static {
+        Properties props = new Properties();
+        try (InputStream is = RemotingVersionInfo.class.getResourceAsStream(RESOURCE_NAME)) {
+            if(is!=null) {
```

Comment:
No need to check for null. If the file is missing, core is corrupt, so die now.

Reply:
The implementation is similar to `jenkins.model.Jenkins`'s appoarch to reading `jenkins-version.properties`. I agree this should not be a problem, but there may be edge cases (e.g. custom packaging). I see no problem with having a fault-tolerant code since it is a regression fix

Reply:
In the case of custom packaging, it is the responsibility of the packager to not make mistakes. If they do, their WAR will be broken. We should not engineer in robustness for something that is not a user error and is trivially detected.

Reply:
Hmm, another potential use-case is calling this class from the agent. I doubt we would be dare enough to send Jenkins core over Remoting. Do we do that in the case of such requests? I hope not

Reply:
Let's keep it as is. I will improve checks when exposing it to public API